File: queue_stack.py
"""
Time Complexity: O(n)
Space compexity: O(n)
"""


class MyQueue:
    def __init__(self):
        self.instack = []
        self.outstack = []

    def intooutstack(self):
        if len(self.instack) != 0:
            i = len(self.instack) - 1

            while(i >= 0):
                self.outstack.append(self.instack[i])
            
                i -= 1   
            self.instack = []

    def push(self, x: int) -> None:
        self.instack.append(x)
        

    def pop(self) -> int:
        if(len(self.outstack) == 0):
            self.intooutstack()
        return self.outstack.pop()
        

    def peek(self) -> int:
        if len(self.outstack) == 0:
            self.intooutstack() 
        return self.outstack[-1]      

    def empty(self) -> bool:
        if len(self.instack) == 0 and len(self.outstack) == 0:
            return True
        return False
        

# Your MyQueue object will be instantiated and called as such:
obj = MyQueue()
obj.push(1)
obj.pop()
obj.peek()
print(obj.empty())

File: queuestack.java
// Time Complexity : O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
/*
implemented the code by creating two stacks instack and outstack and pushing the elements to outstack when pop was needed making th outack a queue */

class MyQueue {
    Stack<Integer> inst;
    Stack<Integer> outst;
    public MyQueue() {
        this.inst = new Stack<>();
        this.outst = new Stack<>();
    }
    
    public void push(int x) {
        inst.push(x);
    }
    
    public int pop() {
        if(outst.empty())
        {
            while(!inst.empty())
            {
                outst.push(inst.pop());
            }
        }
        return outst.pop();
    }
    
    public int peek() {
        if(outst.empty())
        {
            while(!inst.empty())
            {
                outst.push(inst.pop());
            }
        }
        return outst.peek();
    }
    
    public boolean empty() {
        return (outst.empty() && inst.empty());
    }
}